predicatebuilder.key().get(fieldname) not working

have reproduced the problem with the following code:

import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.core.IMap;
import com.hazelcast.query.EntryObject;
import com.hazelcast.query.Predicate;
import com.hazelcast.query.PredicateBuilder;

import java.io.Serializable;
import java.util.Collection;

public class Main {
public static void main(String[] args){
HazelcastInstance hz = Hazelcast.newHazelcastInstance();

```
EntryObject e = new PredicateBuilder().getEntryObject();

Predicate predicate = e.get("priority").greaterEqual(0)
        .and(e.key().get("id").equal("10"));

IMap<LookupKey, LookupValue> hazelcastLookupMap =  hz.getMap("foo");

LookupKey key = new LookupKey();
key.id="10";

LookupValue value = new LookupValue();
value.priority=0;

hazelcastLookupMap.put(key,value);
Collection<LookupValue> result = hazelcastLookupMap.values(predicate);
System.out.println("result.size:"+result.size());
```

}

public static class  LookupKey implements Serializable{
    private String id;

```
public String getId() {
    return id;
}
```

}

public static class LookupValue implements Serializable{
    private long priority;

}
}

And it throws the folowing exception:

Oct 22, 2013 5:00:24 PM com.hazelcast.map.operation.QueryPartitionOperation
SEVERE: [192.168.2.5]:6701 [management-center] com.hazelcast.query.impl.QueryException: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
com.hazelcast.query.impl.QueryException: com.hazelcast.query.impl.QueryException: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
at com.hazelcast.query.impl.QueryEntry.extractViaReflection(QueryEntry.java:88)
at com.hazelcast.query.impl.QueryEntry.getAttribute(QueryEntry.java:78)
at com.hazelcast.query.Predicates$AbstractPredicate.readAttribute(Predicates.java:637)
at com.hazelcast.query.Predicates$EqualPredicate.apply(Predicates.java:563)
at com.hazelcast.query.Predicates$AndPredicate.apply(Predicates.java:337)
at com.hazelcast.query.PredicateBuilder.apply(PredicateBuilder.java:36)
at com.hazelcast.map.MapService.queryOnPartition(MapService.java:919)
at com.hazelcast.map.operation.QueryPartitionOperation.run(QueryPartitionOperation.java:41)
at com.hazelcast.spi.impl.OperationServiceImpl.doRunOperation(OperationServiceImpl.java:274)
at com.hazelcast.spi.impl.OperationServiceImpl.access$800(OperationServiceImpl.java:52)
at com.hazelcast.spi.impl.OperationServiceImpl$LocalOperationProcessor.run(OperationServiceImpl.java:710)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:724)
at com.hazelcast.spi.impl.OperationServiceImpl$OperationThread.run(OperationServiceImpl.java:791)
Caused by: com.hazelcast.query.impl.QueryException: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
at com.hazelcast.query.impl.ReflectionHelper.createGetter(ReflectionHelper.java:138)
at com.hazelcast.query.impl.ReflectionHelper.extractValue(ReflectionHelper.java:143)
at com.hazelcast.query.impl.QueryEntry.extractViaReflection(QueryEntry.java:86)
... 14 more
Caused by: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
at com.hazelcast.query.impl.ReflectionHelper.createGetter(ReflectionHelper.java:128)
... 16 more

Exception in thread "main" com.hazelcast.query.impl.QueryException: com.hazelcast.query.impl.QueryException: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
at com.hazelcast.query.impl.QueryEntry.extractViaReflection(QueryEntry.java:88)
at com.hazelcast.query.impl.QueryEntry.getAttribute(QueryEntry.java:78)
at com.hazelcast.query.Predicates$AbstractPredicate.readAttribute(Predicates.java:637)
at com.hazelcast.query.Predicates$EqualPredicate.apply(Predicates.java:563)
at com.hazelcast.query.Predicates$AndPredicate.apply(Predicates.java:337)
at com.hazelcast.query.PredicateBuilder.apply(PredicateBuilder.java:36)
at com.hazelcast.map.MapService.queryOnPartition(MapService.java:919)
at com.hazelcast.map.operation.QueryPartitionOperation.run(QueryPartitionOperation.java:41)
at com.hazelcast.spi.impl.OperationServiceImpl.doRunOperation(OperationServiceImpl.java:274)
at com.hazelcast.spi.impl.OperationServiceImpl.access$800(OperationServiceImpl.java:52)
at com.hazelcast.spi.impl.OperationServiceImpl$LocalOperationProcessor.run(OperationServiceImpl.java:710)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:724)
at com.hazelcast.spi.impl.OperationServiceImpl$OperationThread.run(OperationServiceImpl.java:791)
Caused by: com.hazelcast.query.impl.QueryException: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
at com.hazelcast.query.impl.ReflectionHelper.createGetter(ReflectionHelper.java:138)
at com.hazelcast.query.impl.ReflectionHelper.extractValue(ReflectionHelper.java:143)
at com.hazelcast.query.impl.QueryEntry.extractViaReflection(QueryEntry.java:86)
... 14 more
Caused by: java.lang.IllegalArgumentException: There is no suitable accessor for 'cokisId'
at com.hazelcast.query.impl.ReflectionHelper.createGetter(ReflectionHelper.java:128)
... 16 more

--- a/hazelcast/src/main/java/com/hazelcast/query/EntryObject.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/EntryObject.java
@@ -16,6 +16,8 @@
 
 package com.hazelcast.query;
 
+import com.hazelcast.query.impl.QueryEntry;
+
 public class EntryObject {
     PredicateBuilder qb;
 
@@ -24,7 +26,21 @@ public class EntryObject {
     }
 
     public EntryObject get(String attribute) {
-        qb.attribute = attribute;
+        if(QueryEntry.KEY_ATTRIBUTE_NAME.equals(qb.attribute)){
+            qb.attribute = QueryEntry.KEY_ATTRIBUTE_NAME+"#"+attribute;
+        }else{
+            qb.attribute = attribute;
+        }
+        return this;
+    }
+
+    public EntryObject key() {
+        if(qb.attribute == null){
+            qb.attribute=QueryEntry.KEY_ATTRIBUTE_NAME;
+        }else if(!qb.attribute.startsWith(QueryEntry.KEY_ATTRIBUTE_NAME)){
+            qb.attribute = QueryEntry.KEY_ATTRIBUTE_NAME+"#"+qb.attribute;
+        }
+
         return this;
     }
 
@@ -34,11 +50,6 @@ public class EntryObject {
 
     public PredicateBuilder isNot(String attribute) {
         return addPredicate(Predicates.notEqual(attribute, true));
-    }
-
-    public EntryObject key() {
-        qb.attribute = "__key";
-        return this;
     }
 
     public PredicateBuilder equal(Comparable value) {

--- a/hazelcast/src/main/java/com/hazelcast/query/impl/QueryEntry.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/impl/QueryEntry.java
@@ -23,7 +23,7 @@ import java.io.IOException;
 public class QueryEntry implements QueryableEntry {
 
     private static final PortableExtractor extractor = new PortableExtractor();
-    private static final String KEY_ATTRIBUTE_NAME = "__key";
+    public static final String KEY_ATTRIBUTE_NAME = "__key";
     private static final String THIS_ATTRIBUTE_NAME = "this";
 
     private final SerializationService serializationService;
@@ -80,10 +80,17 @@ public class QueryEntry implements QueryableEntry {
 
     private Comparable extractViaReflection(String attributeName) {
         try {
-            Object v = getValue();
-            if (KEY_ATTRIBUTE_NAME.equals(attributeName)) return (Comparable) getKey();
-            else if (THIS_ATTRIBUTE_NAME.equals(attributeName)) return (Comparable) v;
-            return ReflectionHelper.extractValue(this, attributeName, v);
+            if (KEY_ATTRIBUTE_NAME.equals(attributeName)) {
+                return (Comparable) getKey();
+            } else if (THIS_ATTRIBUTE_NAME.equals(attributeName)) {
+                return (Comparable) getValue();
+            }
+
+            if(attributeName.startsWith(KEY_ATTRIBUTE_NAME)){
+                return ReflectionHelper.extractValue(this, attributeName, getKey());
+            }else{
+                return ReflectionHelper.extractValue(this, attributeName, getValue());
+            }
         } catch (Exception e) {
             throw new QueryException(e);
         }
@@ -112,6 +119,7 @@ public class QueryEntry implements QueryableEntry {
                 return ReflectionHelper.getAttributeType(this, attributeName);
             }
         }
+
         return ReflectionHelper.getAttributeType(klass);
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java
+++ b/hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java
@@ -73,7 +73,16 @@ public class ReflectionHelper {
     }
 
     private static Getter createGetter(QueryableEntry entry, String attribute) {
-        Object obj = entry.getValue();
+        Object obj;
+        if(attribute.startsWith(QueryEntry.KEY_ATTRIBUTE_NAME)){
+           obj = entry.getKey();
+           if(attribute.length()>QueryEntry.KEY_ATTRIBUTE_NAME.length()){
+               attribute = attribute.substring(QueryEntry.KEY_ATTRIBUTE_NAME.length()+1);
+           }
+        }else{
+           obj = entry.getValue();
+        }
+
         Class clazz = obj.getClass();
         final String cacheKey = clazz.getName() + ":" + attribute;
         Getter getter = getterCache.get(cacheKey);
@@ -125,7 +134,7 @@ public class ReflectionHelper {
                     }
                 }
                 if (localGetter == null) {
-                    throw new IllegalArgumentException("There is no suitable accessor for '" + name + "'");
+                    throw new IllegalArgumentException("There is no suitable accessor for '" + name + "' on class '"+clazz+"'");
                 }
                 parent = localGetter;
             }

--- a/hazelcast/src/test/java/com/hazelcast/query/PredicateBuilderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/query/PredicateBuilderTest.java
@@ -1,0 +1,129 @@
+package com.hazelcast.query;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.test.HazelcastJUnit4ClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.io.Serializable;
+import java.util.Collection;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastJUnit4ClassRunner.class)
+@Category(ParallelTest.class)
+public class PredicateBuilderTest extends HazelcastTestSupport {
+
+    @Test
+    public void get_keyAttribute() {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        HazelcastInstance hz = factory.newHazelcastInstance();
+
+        EntryObject e = new PredicateBuilder().getEntryObject();
+
+        Predicate predicate = e.key().get("id").equal("10");
+
+        IMap<Id, Integer> hazelcastLookupMap = hz.getMap("somemap");
+
+        hazelcastLookupMap.put(new Id("10"), 1);
+        hazelcastLookupMap.put(new Id("20"), 2);
+        hazelcastLookupMap.put(new Id("30"), 3);
+
+        Collection<Integer> result = hazelcastLookupMap.values(predicate);
+        assertEquals(1, result.size());
+        assertTrue(result.contains(1));
+    }
+
+    @Test
+    public void get_key() {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        HazelcastInstance hz = factory.newHazelcastInstance();
+
+        EntryObject e = new PredicateBuilder().getEntryObject();
+
+        Predicate predicate = e.key().equal(10L);
+
+        IMap<Integer, Integer> hazelcastLookupMap = hz.getMap("somemap");
+
+        hazelcastLookupMap.put(10, 1);
+        hazelcastLookupMap.put(30, 2);
+
+        Collection<Integer> result = hazelcastLookupMap.values(predicate);
+        assertEquals(1, result.size());
+        assertTrue(result.contains(1));
+    }
+
+    @Test
+    public void get_this() {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        HazelcastInstance hz = factory.newHazelcastInstance();
+
+        EntryObject e = new PredicateBuilder().getEntryObject();
+
+        Predicate predicate = e.get("this").equal(1L);
+
+        IMap<Integer, Integer> hazelcastLookupMap = hz.getMap("somemap");
+
+        hazelcastLookupMap.put(10, 1);
+        hazelcastLookupMap.put(30, 2);
+
+        Collection<Integer> result = hazelcastLookupMap.values(predicate);
+        assertEquals(1, result.size());
+        assertTrue(result.contains(1));
+    }
+
+    @Test
+    public void get_attribute() {
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        HazelcastInstance hz = factory.newHazelcastInstance();
+
+        EntryObject e = new PredicateBuilder().getEntryObject();
+
+        Predicate predicate = e.get("id").equal("10");
+
+        IMap<Integer, Id> hazelcastLookupMap = hz.getMap("somemap");
+
+        hazelcastLookupMap.put(1,new Id("10"));
+        hazelcastLookupMap.put(2,new Id("20"));
+        hazelcastLookupMap.put(3,new Id("30"));
+
+        Collection<Id> result = hazelcastLookupMap.values(predicate);
+        assertEquals(1, result.size());
+        assertTrue(result.contains(new Id("10")));
+    }
+
+    public static class Id implements Serializable {
+        private String id;
+
+        public Id(String id) {
+            this.id = id;
+        }
+
+        public String getId() {
+            return id;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            Id id1 = (Id) o;
+
+            if (id != null ? !id.equals(id1.id) : id1.id != null) return false;
+
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            return id != null ? id.hashCode() : 0;
+        }
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/test/TestHazelcastInstanceFactory.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/TestHazelcastInstanceFactory.java
@@ -55,6 +55,10 @@ public final class TestHazelcastInstanceFactory {
         }
     }
 
+    public HazelcastInstance newHazelcastInstance() {
+        return newHazelcastInstance(new Config());
+    }
+
     public HazelcastInstance newHazelcastInstance(Config config) {
         if (mockNetwork) {
             if (nodeIndex.get() >= addresses.length) {

